scylla/src/transport/timestamp_generator.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    sync::atomic::AtomicI64,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use std::sync::atomic::Ordering;
+use std::sync::Mutex;
+use tokio::time::{Duration, Instant};
+use tracing::warn;
+
+/// Trait used to represent a timestamp generator
+pub trait TimestampGenerator: Send + Sync {
+    // This generates a new timestamp
+    fn next_timestamp(&self) -> i64;
+}
+
+/// Basic timestamp generator. Provides no guarantees, if system clock returns
+/// time before UNIX epoch it panics.
+#[derive(Default)]
+pub struct SimpleTimestampGenerator {}
+
+#[allow(dead_code)]
+impl SimpleTimestampGenerator {
+    pub fn new() -> Self {
+        SimpleTimestampGenerator {}
+    }
+}
+
+impl TimestampGenerator for SimpleTimestampGenerator {
+    fn next_timestamp(&self) -> i64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_micros() as i64
+    }
+}
+
+/// Warning configuration for MonotonicTimestampGenerator
+pub struct MonotonicTimestampGeneratorCfg {
+    warning_threshold: Duration,
+    warning_interval: Duration,
+}
+
+/// Monotonic timestamp generator. Guarantees monotonicity of timestamps.
+/// If system clock will not provide an increased timestamp, then the timestamp will
+/// be artificially increased. If the config is provided and the clock skew is bigger than
+/// warning_threshold (by default 1 second), then the user will be warned about
+/// the skew repeatedly, with warning_interval provided in the settings (by default 1 second).
+/// Remember that this generator only guarantees monotonicity within one client!
+/// If you create multiple clients with MonotonicTimestampGenerators the monotonicity
+/// guarantee becomes void.
+pub struct MonotonicTimestampGenerator {
+    last: AtomicI64,
+    last_warning: Mutex<Instant>,
+    config: Option<MonotonicTimestampGeneratorCfg>,
+}
+
+impl MonotonicTimestampGenerator {
+    /// Creates a new monotonic timestamp generator with provided settings
+    pub fn new_with_settings(config: Option<MonotonicTimestampGeneratorCfg>) -> Self {
+        MonotonicTimestampGenerator {
+            last: AtomicI64::new(0),
+            last_warning: Mutex::new(Instant::now()),
+            config,
+        }
```

Comment:
This should be done iff we choose to not expose `MonotonicTimestampGeneratorCfg` and instead accept two durations only.

Reply:
In that case, would we like to have the `::new()` version not warning the users?

Reply:
Let's go with non-public MonotonicTimestampGeneratorWarningsCfg, keep `new()` with current config, and expose `new_with_warning_config` (or some other fitting name) accepting two durations.

Reply:
But then we would have the same issue that we started with, the user has to remember to not provide only one of the durations.

Reply:
Right, because the args have to be `Option`'s, otherwise it is not possible to disable warnings, forgot about that.
We could expose third constructor (which disables warnings) or expose the `MonotonicTimestampGeneratorWarnCfg` struct - I'm not a fan of either of the options. @wprzytula ?